drivers/dma/dma_silabs_siwx917.c

```diff
@@ -12,18 +12,26 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/types.h>
 #include "rsi_rom_udma_wrapper.h"
+#include "rsi_rom_udma.h"
 #include "rsi_udma.h"
 #include "sl_status.h"
 
-#define DT_DRV_COMPAT          silabs_siwx917_dma
-#define DMA_MAX_TRANSFER_COUNT 1024
-#define DMA_CH_PRIORITY_HIGH   1
-#define DMA_CH_PRIORITY_LOW    0
-#define VALID_BURST_LENGTH     0
-#define UDMA_ADDR_INC_NONE     0X03
+#define DT_DRV_COMPAT                    silabs_siwx917_dma
+#define DMA_MAX_TRANSFER_COUNT           1024
+#define DMA_CH_PRIORITY_HIGH             1
+#define DMA_CH_PRIORITY_LOW              0
+#define VALID_BURST_LENGTH               0
+#define UDMA_ADDR_INC_NONE               0X03
+#define UDMA_MODE_PER_ALT_SCATTER_GATHER 0x07
 
 LOG_MODULE_REGISTER(si91x_dma, CONFIG_DMA_LOG_LEVEL);
 
+struct dma_sg_descriptor_allocator {
+	/* DMA descriptors in contiguous memory */
+	RSI_UDMA_DESC_T sg_transfer_desc_table[CONFIG_SG_BUFFER_COUNT];
+	uint8_t free_desc[CONFIG_SG_BUFFER_COUNT]; /* Allocation status of descriptors */
```

Comment:
You seem to use the elements as a boolean, so `bool` would be more appropriate than `uint8_t`. However, in this case I'd suggest to go with a bitmap instead. Either `atomic_t` (`ATOMIC_DEFINE()`) or `sys_bitarray_t` (`SYS_BITARRAY_DEFINE())  could be potential candidates here. See e.g. https://docs.zephyrproject.org/latest/doxygen/html/group__bitarray__apis.html

Based on your use case, it seems like `sys_bitarray_t` might be a better fit, letting you e.g. get rid of some of the for-loops you have.

Reply:
Thank you for the suggestion; it helped eliminate some for loops.